File: Python/maxProduct.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: maxProduct.py
@Function: 最大单词长度乘积
@Link: https://leetcode-cn.com/problems/maximum-product-of-word-lengths/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-21
"""


from collections import defaultdict
from typing import List


class Solution:
    def maxProduct(self, words: List[str]) -> int:
        hashmap = defaultdict(int)
        def bit_number(ch): return ord(ch) - ord('a')

        for word in words:
            bitmask = 0
            for ch in word:
                # add bit number bit_number in bitmask
                bitmask |= 1 << bit_number(ch)
            # there could be different words with the same bitmask
            # ex. ab and aabb
            hashmap[bitmask] = max(hashmap[bitmask], len(word))

        max_prod = 0
        for x in hashmap:
            for y in hashmap:
                if x & y == 0:
                    max_prod = max(max_prod, hashmap[x] * hashmap[y])
        return max_prod


# -------------------------
if __name__ == "__main__":
    words = ["abcw", "baz", "foo", "bar", "xtfn", "abcdef"]

    solution = Solution()
    max_product = solution.maxProduct(words)
    print(f"The solution of this problem is {max_product}")

File: C++/maxProduct.cpp
/**
 * @File    : maxProduct.cpp
 * @Brief   : 最大单词长度乘积
 * @Link    : https://leetcode-cn.com/problems/maximum-product-of-word-lengths/
 * @Author  : Wei Li
 * @Date    : 2021-07-21
*/

#include <iostream>
#include <algorithm>
#include <string>
#include <vector>
#include <unordered_map>

class Solution
{
public:
    int maxProduct(std::vector<std::string> &words)
    {
        std::unordered_map<int, int> hash;
        int ans = 0;
        for (const std::string &word : words)
        {
            int mask = 0, size = word.size();
            for (const char &c : word)
            {
                mask |= 1 << (c - 'a');
            }
            hash[mask] = std::max(hash[mask], size);
            // for (const auto &[h_mask, h_len] : hash) // c++ 17 结构化绑定
            for (const auto &ha : hash)
            {
                const auto h_mask = ha.first;
                const auto h_len = ha.second;
                if (!(mask & h_mask))
                {
                    ans = std::max(ans, size * h_len);
                }
            }
        }
        return ans;
    }
};

// -----------------------------
int main(int argc, char **argv)
{
    std::vector<std::string> words{"abcw", "baz", "foo", "bar", "xtfn", "abcdef"};

    Solution solution;
    auto max_product = solution.maxProduct(words);
    std::cout << "The solutio of this problem is " << max_product << std::endl;

    return 0;
}
